sources/packages/backend/libs/test-utils/src/factories/student-restriction.ts

```diff
@@ -0,0 +1,62 @@
+import {
+  Application,
+  Restriction,
+  RestrictionActionType,
+  RestrictionNotificationType,
+  RestrictionType,
+  Student,
+  StudentRestriction,
+  User,
+} from "@sims/sims-db";
+import { createFakeStudent } from "./student";
+import { DataSource } from "typeorm";
+
+/**
+ * Create and save fake student restriction.
+ * @param dataSource data source to persist student restriction.
+ * @param relations student restriction entity relations.
+ * - `student` related student.
+ * - `application` application associated with the student.
+ * - `creator` related user relation.
+ * @param options related to StudentRestriction.
+ * - `restrictionType` option for specifying the restriction type.
+ * - `restrictionCode` option for specifying the restriction code.
+ * - `restrictionCategory` option for specifying the restriction category.
+ * - `actionType` option for specifying the restriction action type.
+ * - `notificationType` option for specifying the restriction notification type.
+ * @returns persisted student restriction.
+ */
+export async function saveFakeStudentRestriction(
+  dataSource: DataSource,
+  relations?: {
+    student?: Student;
+    application?: Application;
+    creator?: User;
+  },
+  options?: {
```

Comment:
The options here are used to filter the restriction but they potentially receive other properties not related to the `where` filter.
Is it not better to follow the pattern from the other helpers that just receive the related entity as a parameter?

Reply:
Do you mean that I should change the `options` to just have the `restrictionCategory` and `notificationType` and accordingly change the `where` condition as well?

Reply:
@sh16011993 not at all. Can we change this to just have the restriction added as a `relations`, as shown below?
```ts
export async function saveFakeStudentRestriction(
  dataSource: DataSource,
  relations: {
    restriction: Restriction;
    student?: Student;
    application?: Application;
    creator?: User;
  },
): Promise<StudentRestriction> {
  const studentRestrictionRepo = dataSource.getRepository(StudentRestriction);
  const studentRestriction = new StudentRestriction();
  studentRestriction.student = relations?.student ?? createFakeStudent();
  studentRestriction.application = relations?.application;
  studentRestriction.isActive = true;
  studentRestriction.restriction = relations.restriction;
  studentRestriction.creator = relations?.creator;
  return studentRestrictionRepo.save(studentRestriction);
}
```

In the consumer part it would be:
```ts
const restriction = await db.restriction.findOne({ ... });
const studentRestriction = await saveFakeStudentRestriction(appDataSource, {
  restriction,
  student,
  application,
  creator: student.user
});
```

The second option suggested was:
```ts
export async function saveFakeStudentRestriction(
  dataSource: DataSource,
  relations?: {
    student?: Student;
    application?: Application;
    creator?: User;
  },
  options?: {
    restrictionFilter?: FindOptionsWhere<Restriction>;
  },
): Promise<StudentRestriction> {
  const restrictionRepo = dataSource.getRepository(Restriction);
  const restriction = await restrictionRepo.findOne({
    where: options?.restrictionFilter,
  });
  const studentRestrictionRepo = dataSource.getRepository(StudentRestriction);
  const studentRestriction = new StudentRestriction();
  studentRestriction.student = relations?.student ?? createFakeStudent();
  studentRestriction.application = relations?.application;
  studentRestriction.isActive = true;
  studentRestriction.restriction = restriction;
  studentRestriction.creator = relations?.creator;
  return studentRestrictionRepo.save(studentRestriction);
}
```

Reply:
Thank You Andrew.

Reply:
I liked both approaches. The first one is more in sync with what we have now in our system.